src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs

```diff
@@ -199,6 +204,12 @@ bool IsValidTriggerCharacterForDocument(Document document, char triggerCharacter
                     }
                 }
 
+                if (commitCharacters.SetEquals(CompletionRules.Default.DefaultCommitCharacters))
+                {
+                    currentRuleCache.Add(item.Rules.CommitCharacterRules, ImmutableArray<string>.Empty);
+                    return null;
```

Comment:
good catch, this will definitely save a bunch.

Reply:
this seems like a bunch of computation and allocation per item.  do we know how impactful that is?

Reply:
I would imagine there are only a few different sets of rules per completion list, this occurs after the cache check step and caches its own result.

Reply:
can you doc this as well?  i don't quite understand what's going on.  why are we adding 'empty' to the rule cache?  does tihs mean the current item has no commit characters?

Reply:
I will add a comment. It means that the current item only has default commit characters, so we cache Empty as a null equivalent to avoid doing the default character check in the future.